### ISSUE TITLE
Support non-Debizium Update/Delete data

### ISSUE THREAD

**Issue by gregberns** [MaterializeInc/materialize#2125]
Congrats on the release it looks great and the documentation is very good.

We have a legacy database not supported by Debezium (and never will be, its very legacy), and I'd like to get CDC data from it and send it to Materialized. Update/Delete support is not currently supported without using Debezium. This request is to get 'update/delete' support for data NOT coming from Debezium - meaning sending hand build messages, where the message schema is defined in the "CREATE SOURCE" command.

I've been attempting to reverse engineer how Materialized is reading the Debezium data from Kafka, but I'm missing something. When running `demo/simple` and looking at messages in the `mysql.inventory.products` topic with `kafkacat`(`kafkacat -C -b kafka:9092 -t mysql.inventory.products -J  -o -1 -e`) the messages don't appear to contain the `before` and `after` fields. Maybe this is the job of the Kafka Connect tool?

```
{"topic":"mysql.inventory.products","partition":0,"offset":8,"key":"\u0000\u0000\u0000\u0000\u000B?\u0001","payload":"\u0000\u0000\u0000\u0000\f\u0000\u0002?\u0001\u0014spare tire\u0002$24 inch spare tire\u0002\u0000\u0000\u0000@336@\u00161.0.1.Final\nmysql\nmysql\u0000\u0000\btrue\u0012inventory\u0002\u0010products\u0000\u0000 mysql-bin.000003??$\u0000\u0000\u0000\u0002c\u0002ʁ?܎\\"}
```

## Objective

Add messages to Kafka with a well defined format and create a Materialized view off the topic. Based on the data supplied, a message can update/delete the data from a previous message.

Details:

* Message may need a 'key', like table + primary key thats defined in the payload
* Format would be an Avro message with `before`/`after` fields
* `before`/`after` could be a Union and list of `(KEY, VALUE)` tuples
* Update/Insert/Delete would be done by:
  * Insert: before is `None` and after is `Some`
  * Update: before is `Some` and after is `Some`
  * Delete: before is `Some` and after is `None`

It appears that the current Debezium record format is something like this Rust snippet.

```
let mut record = Value::Record(vec![
        (
            "before".to_string(),
            Value::Union(Box::new(Value::Record(vec![
                ("FirstName".to_string(), Value::String("Pink".to_string())),
                (
                    "LastName".to_string(),
                    Value::String("Elephants".to_string()),
                ),
            ]))),
        ),
        (
            "after".to_string(),
            Value::Union(Box::new(Value::Record(vec![
                (
                    "FirstName".to_string(),
                    Value::String("Donnatella".to_string()),
                ),
                ("LastName".to_string(), Value::String("Moss".to_string())),
            ]))),
        ),
    ]);
```

## Current Problems

I haven't been able to figure out how to serialize the messages in Avro so that Materialized can consume them. (I can put up an example repo of what I've tried if thats helpful)

* Is the whole message need to be in Avro?
* Or is it JSON (from Kafka Connect) with Avro in the `payload` field?

I've looked through `dataflow/decode/avro.rs` and `interchange/avro.rs` and I can seem to get past errors that would be thrown at the [top of `decode`](https://github.com/MaterializeInc/materialize/blob/master/src/interchange/avro.rs#L394), but when I setup the materialized view the query will fail with `column \"{}\" does not exist`.

## Next Steps

First, any guidance on how this might be accomplished with little to no update to the current code base (but still remaining unsupported?)
Is this support on the roadmap in the near future? (next 3 months)
If not, is there anything I can do to help facilitate?

**Comment by gregberns** [comment]
Very similar to #652 and #207.

**Comment by ruchirK** [comment]
Hi Greg! Thank you for your interest in using Materialize! It sounds like you are trying to hand define a source with a inline Avro schema that matches the before/after diff envelope format that we get from Debezium and use that to ingest CDC data from your database -- my understanding is accurate right?

If so, then as far as I can tell -- that should work / your snippet of Rust code describing a message in the currently supported format looks right. To your other question - the whole message needs to be in Avro at the moment. 

Any further details / concrete steps to reproduce the error that you are seeing would be extremely useful. Thanks!

**Comment by gregberns** [comment]
Here's an [example repository](https://github.com/gregberns/materialize-writer). Steps are documented in the readme.

## Errors

In the example code, there are two attempts at passing a message.

To run the first example, go to `main()` and uncomment line 102-103 and comment 106-107.

### Kafka Message with Schema

`src/main.rs:serialize()` will use the [`MaterializeInc/avro_rs`](https://github.com/MaterializeInc/avro-rs/) version of the Avro `Writer`.

When Materialized processes this message it fails with the error at the [top of `decode`](https://github.com/MaterializeInc/materialize/blob/master/src/interchange/avro.rs#L394).

When I looked at the message in Kafka, the Avro `Writer` is putting the whole schema in the payload.

I also tried to add the 'magic' header manually to the message (see `serialize()`), but that didnt seem to help.

### Kafka Message with Without Schema

So instead of using the `Writer` directly I wanted to encode the `Value::Record` and send it, as seen in `src/main.rs:serialize2()`.

To do this, I extracted the private [MaterializeInc/avro_rs `encode()` method](https://github.com/MaterializeInc/avro-rs/blob/master/src/encode.rs) into `src/avro_encode.rs` to test it out.

When I run the workflow below, I get all the way to the last step which fails with `column "FirstName" does not exist`, which you can find in the Materialized source code as `column \"{}\" does not exist`.

**Comment by umanwizard** [comment]
Hi @gregberns -- I have attempted to run your demo. Regardless of whether I run it with lines 102-103 or 106-107, I am getting the "wrong avro serialization magic" error.

On this line:

```
    let mut body = [b'O', b'1', b'1', b'1', b'1'].to_vec();
```

you are setting the magic byte to the ASCII character 'O' (0x45), rather than a null byte (0x00) as required by https://docs.confluent.io/current/schema-registry/docs/serializer-formatter.html#wire-format and expected by Materialize.

**Comment by benesch** [comment]
@umanwizard, food for thought: The magic header really only makes sense if
folks are using the CSR. Probably inline and file Avro schemas should
create a source that does not expect the header.

**Comment by umanwizard** [comment]
@gregberns Changing the magic number to 0, I am now getting the following behavior:

```
materialize=> create materialized view v as select * from topic1;
CREATE VIEW
materialize=> select * from v;
ERROR:  canceling statement due to user request
```

This is probably a bug in materialize

**Comment by gregberns** [comment]
> you are setting the magic byte to the ASCII character 'O' (0x45), rather than a null byte (0x00)

Of course! I thought I set it to zero, but I'm an idiot makes more sense that its 0x00.

What is "CSR"?:  "The magic header really only makes sense if folks are using the CSR."

Thanks for looking into it!

**Comment by umanwizard** [comment]
@gregberns "CSR" is Confluent Schema Registry -- @benesch was pointing out that when you're not using the CSR, it would make sense for Materialize to support Avro streams without this magic header.

**Comment by quodlibetor** [comment]
I believe that if someone wanted to do this today, with their own CDC format, the general guidance would be to ingest your data into two views and then create a materialized view that is all of the inserted rows, except all the removed rows.

So if you had a wire format that looks like:

```javascript
{
    "name": "cdc",
    "type": "record",
    "fields": [
        {
          "name": "removed",
          "type": "record",
          "fields": [
             "name": "text",
             "type": "string"
          ]
        },
        {
          "name": "inserted",
          "type": "record",
          "fields": [
             "name": "text",
             "type": "string"
          ]
        }
    ]
}
```

(note that it is `removed`/`inserted` instead of `before`/`after`, but that otherwise all the fields are the same).

You could create your own source that looks like:

```sql
CREATE SOURCE insertions FROM KAFKA BROKER 'b' TOPIC 't' FORMAT AVRO
    USING SCHEMA '<the-above>' WITH (confluent_wire_format = false);
CREATE SOURCE removals FROM KAFKA BROKER 'b' TOPIC 't' FORMAT AVRO
    USING SCHEMA '<the-above>' WITH (confluent_wire_format = false);
CREATE VIEW insertions_view AS SELECT (cdc).(inserted).text text FROM insertions;
CREATE VIEW removals_view AS SELECT (cdc).(removed).text text FROM insertions;

CREATE MATERIALIZED VIEW cdc AS
    SELECT * FROM insertions_view
    EXCEPT ALL
    SELECT * FROM removals_view;
```

And this will just work, and should be reasonably efficient, but @frankmcsherry can correct me about any of that.
